File: apps/backend/src/registry/registry_test.go
package registry_test

import (
	"os"
	"path/filepath"
	"testing"

	"github.com/lucas/oraculo/apps/backend/src/registry"
)

func TestRegister_CreatesFileAndAddsEntry(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "servers.json")

	err := registry.Register(path, registry.Entry{
		Project: "test-project",
		Path:    "/tmp/test",
		Port:    3100,
		PID:     os.Getpid(),
	})
	if err != nil {
		t.Fatalf("Register: %v", err)
	}

	entries, err := registry.List(path)
	if err != nil {
		t.Fatalf("List: %v", err)
	}
	if len(entries) != 1 {
		t.Fatalf("expected 1 entry, got %d", len(entries))
	}
	if entries[0].Project != "test-project" {
		t.Errorf("expected project 'test-project', got %q", entries[0].Project)
	}
	if entries[0].Port != 3100 {
		t.Errorf("expected port 3100, got %d", entries[0].Port)
	}
}

func TestUnregister_RemovesEntry(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "servers.json")

	_ = registry.Register(path, registry.Entry{Project: "a", Path: "/a", Port: 3100, PID: os.Getpid()})
	_ = registry.Register(path, registry.Entry{Project: "b", Path: "/b", Port: 3101, PID: os.Getpid()})

	err := registry.Unregister(path, "/a")
	if err != nil {
		t.Fatalf("Unregister: %v", err)
	}

	entries, _ := registry.List(path)
	if len(entries) != 1 {
		t.Fatalf("expected 1 entry, got %d", len(entries))
	}
	if entries[0].Path != "/b" {
		t.Errorf("expected /b, got %q", entries[0].Path)
	}
}

func TestRegister_UpdatesExistingEntry(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "servers.json")

	_ = registry.Register(path, registry.Entry{Project: "a", Path: "/a", Port: 3100, PID: 1})
	_ = registry.Register(path, registry.Entry{Project: "a", Path: "/a", Port: 3200, PID: 2})

	entries, _ := registry.List(path)
	if len(entries) != 1 {
		t.Fatalf("expected 1 entry, got %d", len(entries))
	}
	if entries[0].Port != 3200 {
		t.Errorf("expected port 3200, got %d", entries[0].Port)
	}
}

func TestList_ReturnsNilForMissingFile(t *testing.T) {
	entries, err := registry.List("/nonexistent/servers.json")
	if err != nil {
		t.Fatalf("List: %v", err)
	}
	if entries != nil {
		t.Errorf("expected nil, got %v", entries)
	}
}

File: apps/orchestrator/src/registry/__tests__/registry.test.ts
import { describe, it, expect, beforeEach, afterEach } from "bun:test";
import { mkdtemp, rm } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { register, unregister, list, defaultPath } from "../registry.ts";
import type { ServerEntry } from "../registry.ts";

const ts = "2026-01-01T00:00:00.000Z";

describe("registry", () => {
  let tempDir: string;
  let registryPath: string;

  beforeEach(async () => {
    tempDir = await mkdtemp(join(tmpdir(), "oraculo-registry-test-"));
    registryPath = join(tempDir, "servers.json");
  });

  afterEach(async () => {
    await rm(tempDir, { recursive: true, force: true });
  });

  describe("register", () => {
    it("creates registry file and adds entry", async () => {
      const entry: ServerEntry = {
        project: "test-project",
        port: 8080,
        pid: process.pid,
        path: "/tmp/test",
        started_at: ts,
      };

      await register(registryPath, entry);
      const entries = await list(registryPath);
      expect(entries).toHaveLength(1);
      expect(entries[0].project).toBe("test-project");
      expect(entries[0].port).toBe(8080);
    });

    it("updates existing entry with same path", async () => {
      const entry1: ServerEntry = {
        project: "proj",
        port: 8080,
        pid: process.pid,
        path: "/tmp/test",
        started_at: ts,
      };
      await register(registryPath, entry1);

      const entry2: ServerEntry = {
        project: "proj",
        port: 9090,
        pid: process.pid,
        path: "/tmp/test",
        started_at: ts,
      };
      await register(registryPath, entry2);

      const entries = await list(registryPath);
      expect(entries).toHaveLength(1);
      expect(entries[0].port).toBe(9090);
    });

    it("adds multiple entries with different paths", async () => {
      await register(registryPath, {
        project: "proj-a",
        port: 8080,
        pid: process.pid,
        path: "/tmp/a",
        started_at: ts,
      });
      await register(registryPath, {
        project: "proj-b",
        port: 8081,
        pid: process.pid,
        path: "/tmp/b",
        started_at: ts,
      });

      const entries = await list(registryPath);
      expect(entries).toHaveLength(2);
    });

    it("creates parent directories if needed", async () => {
      const nestedPath = join(tempDir, "nested", "dir", "servers.json");
      await register(nestedPath, {
        project: "proj",
        port: 8080,
        pid: process.pid,
        path: "/tmp/test",
        started_at: ts,
      });

      const entries = await list(nestedPath);
      expect(entries).toHaveLength(1);
    });
  });

  describe("unregister", () => {
    it("removes entry by path", async () => {
      await register(registryPath, {
        project: "proj",
        port: 8080,
        pid: process.pid,
        path: "/tmp/test",
        started_at: ts,
      });

      await unregister(registryPath, "/tmp/test");
      const entries = await list(registryPath);
      expect(entries).toHaveLength(0);
    });

    it("does nothing when path not found", async () => {
      await register(registryPath, {
        project: "proj",
        port: 8080,
        pid: process.pid,
        path: "/tmp/test",
        started_at: ts,
      });

      await unregister(registryPath, "/tmp/nonexistent");
      const entries = await list(registryPath);
      expect(entries).toHaveLength(1);
    });

    it("does nothing when registry file does not exist", async () => {
      await expect(unregister(registryPath, "/tmp/test")).resolves.toBeUndefined();
    });
  });

  describe("list", () => {
    it("returns empty array when file does not exist", async () => {
      const entries = await list(registryPath);
      expect(entries).toEqual([]);
    });

    it("filters out stale entries with non-running PIDs", async () => {
      // Write entries directly — one with current PID (alive), one with bogus PID (stale)
      const entries: ServerEntry[] = [
        { project: "alive", port: 8080, pid: process.pid, path: "/tmp/alive", started_at: ts },
        { project: "stale", port: 8081, pid: 999999, path: "/tmp/stale", started_at: ts },
      ];
      await Bun.write(registryPath, JSON.stringify(entries, null, 2) + "\n");

      const result = await list(registryPath);
      expect(result).toHaveLength(1);
      expect(result[0].project).toBe("alive");
    });
  });

  describe("defaultPath", () => {
    it("returns path under home directory", () => {
      const p = defaultPath();
      expect(p).toContain(".oraculo");
      expect(p).toEndWith("servers.json");
    });
  });
});
